index op with int32 support

jianyuh opened the pull request:
Differential Revision: D40089960

malfet commented:
Looks good to me, but just curious why

jianyuh commented:
> Looks good to me, but just curious why

Thanks for approving! See more context in https://github.com/pytorch/pytorch/pull/86309 . I will wait for @ngimel 's signal to check if this change is good.

ngimel commented on aten/src/ATen/native/IndexingUtils.h:
```diff
@@ -52,8 +52,8 @@ static C10_UNUSED void checkIndexTensorTypes(IOptTensorListRef indices) {
       auto scalarType = tensor->scalar_type();
-      if (scalarType != kLong && scalarType != kByte && scalarType != kBool) {
-          TORCH_CHECK_INDEX(false, "tensors used as indices must be long, byte or bool tensors");
+      if (scalarType != kLong && scalarType != kInt && scalarType != kByte && scalarType != kBool) {
```
can you universally expand the types accepted by this function? It's called from many places, some of which (maybe) don't accept int32 indices.

jianyuh replied:
https://github.com/pytorch/pytorch/pull/86309 is more reasonable approach. Was trying to check if we can directly remove int constraint but looks like it fails even for CPU index ops.

ngimel commented:
Wait, but `index_select` already works with int32 today? @qizzzh enabled this some time back.

jianyuh committed:
index op with int32 support (#86318)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/86318

As title.

Test Plan: buck2 run mode/opt  -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c python.package_style=inplace -c fbcode.nvcc_arch=a100 //ai_codesign/nonprod/jianyuhuang/pytorch_examples:int_index_check

Differential Revision: D40089960

fbshipit-source-id: 839d73103ae5fb7deb0de0f98b416d8aaba82a46

jianyuh commented:
> Wait, but `index_select` already works with int32 today? @qizzzh enabled this some time back.

Sorry I meant to check index op.

jianyuh commented:
@pytorchbot merge
